lib/awful/screen/dpi.lua

```diff
@@ -0,0 +1,414 @@
+---------------------------------------------------------------------------
+--                          DPI detection code.                          --
+---------------------------------------------------------------------------
+
+local capi    = {screen = screen}
+local gtable  = require("gears.table")
+local grect   = require("gears.geometry").rectangle
+local gdebug  = require("gears.debug")
+
+local module = {}
+
+local ascreen, data = nil, nil
+
+-- Metric to Imperial conversion constant.
+local mm_per_inch = 25.4
+
+local xft_dpi, fallback_dpi
+
+local function get_fallback_dpi()
+    -- Following Keith Packard's whitepaper on Xft,
+    -- https://keithp.com/~keithp/talks/xtc2001/paper/xft.html#sec-editing
+    -- the proper fallback for Xft.dpi is the vertical DPI reported by
+    -- the X server. This will generally be 96 on Xorg, unless the user
+    -- has configured it differently
+    if not fallback_dpi then
+        local _, h = root.size()
+        local _, hmm = root.size_mm()
+        fallback_dpi = hmm ~= 0 and h * mm_per_inch / hmm
+    end
+
+    return math.floor(fallback_dpi or 96)
```

Comment:
Why are you rounding the DPI down? At least Pango uses `double` for DPI and besides that I do not know which piece of code should care.

Reply:
I am unsure if rounding is a good idea here. The reason I did it was because writing a rule like:

```
-- I declare I like small bars.
ruled.screen.append_rule {
    rule = {dpi = 321.325325435435 },
    properties = {dpi = 128 }
}
```

Fells a bit wrong and error prone. `==` is by default for matching (even if, as you saw, I added a magic way to work around this). So most users will find having to enter long floating point number inconvenient rather than having them rounded.

Yes sure, I added rule ranges for exactly this use case, but still...

What do you think?

Reply:
I think that someone somewhere somewhen will want to have a fractional DPI (I could link to the relevant XKCD if you want) and "but then rules do not work" will not make them happy. And it feels like a bad idea.

Your "I like small bars"-example should use "if DPI is larger than 200, use 128 instead".

If you really want to round DPIs for rule matching, I'd say that then DPIs should be rounded only for rule matching and not globally. I bet that will result in uglier code, but at least conceptually it feels a lot less weird.

Reply:
![image](https://user-images.githubusercontent.com/340384/62420651-ec100080-b663-11e9-8740-3fd890b4666f.png)

Reply:
or mb just allow `gt`, `ge`, `lt`, `le` conditions in the rules :D

Reply:
DPI rounding isn't even that much of an issue, since the DPI value itself is rarely used as-is, and any rounding can (and arguably should) be applied at point-of-use (e.g. scaled lengths and font sizes). Most users would probably care more about the preferred scaling factor, expressable as a ratio of DPIs, and _that_ one could (should?) be rounded.

If we do provide a per-screen scaling_factor / pixel_ratio property (see e.g. Firefox' about:config and the Qt API), then _that_ should be rounded (per the user preference), otherwise it might be better to leave the DPI unrounded, and instead provide functions to round the scaled lenghs.

(The reason why I think that the scaling factor rounding should be user-controlled is that with non-perfectly integer scaling it's entirely up to the user to choose if they want rounding up, down or no rounding at all. Maybe define some kind of rounding policy property? [The Qt issue on the matter](https://bugreports.qt.io/browse/QTBUG-53022) has some potentially interesting ideas here.)